internal/rest/leveldbreceipts.go

```diff
@@ -0,0 +1,373 @@
+// Copyright 2021 Kaleido
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rest
+
+import (
+	"encoding/json"
+	"fmt"
+	"math/rand"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/kaleido-io/ethconnect/internal/errors"
+	"github.com/kaleido-io/ethconnect/internal/kvstore"
+	"github.com/oklog/ulid/v2"
+	log "github.com/sirupsen/logrus"
+	"github.com/syndtr/goleveldb/leveldb/util"
+)
+
+type levelDBReceipts struct {
+	conf         *LevelDBReceiptStoreConf
+	store        kvstore.KVStore
+	idEntropy    *ulid.MonotonicEntropy
+	defaultLimit int
+}
+
+func newLevelDBReceipts(conf *LevelDBReceiptStoreConf) (*levelDBReceipts, error) {
+	store, err := kvstore.NewLDBKeyValueStore(conf.Path)
+	if err != nil {
+		return nil, errors.Errorf(errors.ReceiptStoreLevelDBConnect, err)
+	}
+	t := time.Unix(1000000, 0)
+	entropy := ulid.Monotonic(rand.New(rand.NewSource(t.UnixNano())), 0)
+
+	return &levelDBReceipts{
+		conf:         conf,
+		store:        store,
+		idEntropy:    entropy,
+		defaultLimit: 50,
+	}, nil
+}
+
+// AddReceipt processes an individual reply message, and contains all errors
+// To account for any transitory failures writing to mongoDB, it retries adding receipt with a backoff
+func (l *levelDBReceipts) AddReceipt(requestID string, receipt *map[string]interface{}) (err error) {
+	// insert an entry with a composite key to track the insertion order
+	newId := ulid.MustNew(ulid.Timestamp(time.Now()), l.idEntropy)
+	// add "z" prefix so these entries come after the lookup entries
+	// because for iteration we start from last backwards
+	lookupKey := fmt.Sprintf("z%s", newId)
+
+	b, _ := json.MarshalIndent(receipt, "", "  ")
+	l.store.Put(lookupKey, b)
+
+	// build the index for "from"
+	fromKey := fmt.Sprintf("from:%s:%s", (*receipt)["from"], lookupKey)
```

Comment:
Sorry for the amount of work needed to create the indexed filtering in LevelDB.
I did do a bit of research around how problematic having a huge number of keys is, and people were talking in the low billions. So I don't _think_ having 5 keys per receipt is a problem, and for backwards compatibility I don't think we can avoid it. However, I wonder if for new users (particularly FireFly driven usage) we should have a config option to disable rich searching.

Reply:
good idea.

Reply:
Discussed with @jimthematrix and this is deferred. Performance testing did not show an issues in the millions of transaction range.